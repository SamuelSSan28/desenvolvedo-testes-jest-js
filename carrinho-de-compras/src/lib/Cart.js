import { find, remove } from 'lodash';
import Dinero from 'dinero.js';

const Money = Dinero;
Money.defaultCurrenc = 'BRL';
Money.defaultPrecision = 2;

export default class Cart {
  itens = [];

  add(item) {
    const itemToFind = { product: item.product };

    if (find(this.itens, itemToFind)) {
      remove(this.itens, itemToFind);
    }

    this.itens.push(item);
  }

  remove(product) {
    remove(this.itens, { product });
  }

  calculatePercentageDiscount(amount, item) {
    if (item.condition?.percentage && item.quantity > item.condition.minimum) {
      return amount.percentage(item.condition.percentage);
    }
    return Money({ amount: 0 });
  }

  calculateQuantityDiscount(amount, item) {
    if (item.condition?.quantity && item.quantity > item.condition.quantity) {
      return amount.percentage(50);
    }
    return Money({ amount: 0 });
  }

  calculateDiscount(amount, quantity, condition) {
    const list = Array.isArray(condition) ? condition : [condition];
    const [higherDiscount] = list
      .map((cond) => {
        if (cond.percentage) {
          return this.calculatePercentageDiscount(amount, {
            quantity,
            condition:cond,
          }).getAmount();
        } else if (cond.quantity) {
          return this.calculateQuantityDiscount(amount, {
            quantity,
            condition:cond,
          }).getAmount();
        }
      })
      .sort((a, b) => b-a);
      
      return  Money({amount:higherDiscount})
  }

  getTotal() {
    return this.itens.reduce((acc, item) => {
      const amount = Money({ amount: item.quantity * item.product.price });
      let discount = Money({ amount: 0 });

      if (item.condition) {
        discount = this.calculateDiscount(
          amount,
          item.quantity,
          item.condition
        );
      }
      return acc.add(amount).subtract(discount);
    }, Money({ amount: 0 }));
  }

  sumary() {
    const total = this.getTotal().getAmount();
    const itens = this.itens;

    return {
      total,
      itens,
    };
  }

  checkout() {
    const { total, itens } = this.sumary();

    this.itens = [];

    return {
      total,
      itens,
    };
  }
}
