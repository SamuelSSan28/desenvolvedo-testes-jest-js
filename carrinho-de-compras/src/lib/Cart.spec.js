import Cart from './cart';

let cart;
const product = {
  title: 'CD do Mamonas Assasinas',
  price: 35588,
};

const product2 = {
  title: 'CD do RaÃ§a Negra',
  price: 11111,
};

beforeEach(() => {
  cart = new Cart();
});

describe('getTotal()', () => {
  it('should return 0 when getTotal() is executed in a newly created instance', () => {
    expect(cart.getTotal().getAmount()).toBe(0);
  });

  it('should multiply quantity and price and receive the total amount', () => {
    const item = {
      product: {
        title: 'CD do Mamonas Assasinas',
        price: 35588,
      },
      quantity: 2,
    };

    cart.add(item);
    expect(cart.getTotal().getAmount()).toBe(71176);
  });

  it('should ensure no more then one product exists at a time', () => {
    cart.add({
      product,
      quantity: 2,
    });

    cart.add({
      product,
      quantity: 1,
    });

    expect(cart.getTotal().getAmount()).toBe(35588);
  });

  it('should update total when a product gets included and them remove', () => {
    cart.add({
      product,
      quantity: 1,
    });

    expect(cart.getTotal().getAmount()).toBe(35588);

    cart.add({
      product: product2,
      quantity: 1,
    });

    expect(cart.getTotal().getAmount()).toBe(46699);

    cart.remove(product2);

    expect(cart.getTotal().getAmount()).toBe(35588);
  });
});

describe('checkout()', () => {
  it('should return an object with the total and the list of items', () => {
    cart.add({
      product,
      quantity: 1,
    });

    cart.add({
      product: product2,
      quantity: 1,
    });

    expect(cart.sumary()).toMatchSnapshot();
    expect(cart.getTotal().getAmount()).toBeGreaterThan(0);
  });

  it('should reset the cart when checkout() is called', () => {
    cart.add({
      product,
      quantity: 1,
    });

    cart.checkout();

    expect(cart.getTotal().getAmount()).toEqual(0);
  });
});

describe('special conditions', () => {
  it('should apply percentage discount quantity above minimum is passed ', () => {
    const condition = {
      percentage: 30,
      minimum: 2,
    };

    cart.add({
      product,
      condition,
      quantity: 3,
    });

    expect(cart.getTotal().getAmount()).toEqual(74735);
  });

  it('should not apply percentage discount quantity when is less or equal minimum', () => {
    const condition = {
      percentage: 30,
      minimum: 2,
    };

    cart.add({
      product,
      condition,
      quantity: 2,
    });

    expect(cart.getTotal().getAmount()).toEqual(71176);
  });

  it('should apply percentage discount quantity for even quantities', () => {
    const condition = {
      quantity: 2,
    };

    cart.add({
      product,
      condition,
      quantity: 3,
    });

    expect(cart.getTotal().getAmount()).toEqual(53382);
  });

  it('should receve two or more conditions and apply best discont, first case', () => {
    const condition = {
      quantity: 2,
    };
    const condition2 = {
      percentage: 30,
      minimum: 2,
    };

    cart.add({
      product,
      condition:[condition,condition2],
      quantity: 3,
    });

    expect(cart.getTotal().getAmount()).toEqual(53382);
  });
});
