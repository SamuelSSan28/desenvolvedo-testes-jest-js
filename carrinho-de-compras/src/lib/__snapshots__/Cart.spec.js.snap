// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`checkout() should return an object with the total and the list of items 1`] = `
Object {
  "itens": Array [
    Object {
      "product": Object {
        "price": 35588,
        "title": "CD do Mamonas Assasinas",
      },
      "quantity": 1,
    },
    Object {
      "product": Object {
        "price": 11111,
        "title": "CD do Ra√ßa Negra",
      },
      "quantity": 1,
    },
  ],
  "total": 46699,
}
`;
